### PR TITLE
Dpp 1639 eas changes

### PR DESCRIPTION
**Pull request by carlbm**

### CODE REVIEW

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using SFA.DAS.Payments.Calc.ProviderAdjustments.Domain;
+using SFA.DAS.Payments.Calc.ProviderAdjustments.Domain.Extensions;
+using SFA.DAS.Payments.Calc.ProviderAdjustments.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.Payments.Calc.ProviderAdjustments.Services
+{
+    public interface ICalculateProviderPayments
```
Comment (Pauljgraham) [inline src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs]: We should have a standard approach on where we put interfaces.

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        List<PaymentEntity> CalculatePayments(List<AdjustmentEntity> previousPayments,
+            List<AdjustmentEntity> earnings);
+    }
+
+    public class ProviderPaymentsCalculator : ICalculateProviderPayments
+    {
+        public List<PaymentEntity> CalculatePayments(
```
Comment (Pauljgraham) [inline src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs]: Could we change the name to CalculatePaymentsARefunds or something ....

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Domain/Extensions/ProviderPaymentsGroupExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿using SFA.DAS.Payments.Calc.ProviderAdjustments.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.Payments.Calc.ProviderAdjustments.Domain.Extensions
+{
+    public static class ProviderPaymentsGroupExtensions
+    {
+        public static PaymentEntity MakePayment(this ProviderPaymentsGroup source, decimal amount)
```
Comment (Pauljgraham) [inline src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Domain/Extensions/ProviderPaymentsGroupExtensions.cs]: MakePayment sounds like a processing function, would it be better to call it CreatePaymentEntity. 

Also does it need to be an extension method? Doesn't it belong to the ProviderPaymentsCalculator as a private method?

File: src/PeriodEnd/ProviderAdjustments/DeployProviderAdjustments/sql/dml/06 PeriodEnd.ProviderAdjustments.Populate.Reference.Current.dml.sql [hunk 4]
```diff
@@ -16,28 +10,17 @@ INSERT INTO [Reference].[ProviderAdjustmentsCurrent] (
...
-	s.Ukprn IN (SELECT Ukprn FROM Reference.ProviderAdjustmentsProviders)
-    AND s.Submission_Id NOT IN (SELECT SubmissionId FROM @AlreadyProcessedSubmissions)
+	S.Ukprn,
+	S.Submission_Id [SubmissionId],
+	S.CollectionPeriod [SubmissionCollectionPeriod],
+	V.Payment_Id [PaymentType],
+	P.PaymentName [PaymentTypeName],
+	V.PaymentValue [Amount]
+	
+FROM ${EAS_Deds.FQ}.dbo.EAS_Submission_Values V
```
Comment (Pauljgraham) [inline src/PeriodEnd/ProviderAdjustments/DeployProviderAdjustments/sql/dml/06 PeriodEnd.ProviderAdjustments.Populate.Reference.Current.dml.sql:20]: If OpenQuery was introduced previously for performance reasons, why is it being removed?

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments.UnitTests/Sandbox.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using FluentAssertions;
+using NUnit.Framework;
+
+namespace SFA.DAS.Payments.Calc.ProviderAdjustments.UnitTests
+{
+    [TestFixture]
+    public class Sandbox
+    {
```
Comment (Pauljgraham) [inline src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments.UnitTests/Sandbox.cs:10]: Are these tests needed? As it's not testing our code

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Domain/ProviderPaymentsGroup.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using SFA.DAS.Payments.Calc.ProviderAdjustments.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.Payments.Calc.ProviderAdjustments.Domain
+{
+    public class ProviderPaymentsGroup : IEquatable<ProviderPaymentsGroup>
+    {
+        public long Ukprn { get; }
+        public int PaymentType { get; }
```
Comment (Pauljgraham) [inline src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Domain/ProviderPaymentsGroup.cs:9]: Should this be a enum?

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                var paymentAmount = earningGroup.Sum(x => x.Amount) -
+                                    groupedPreviousPayments[earningGroup.Key].Sum(x => x.Amount);
+
+                if (paymentAmount != 0)
+                {
+                    payments.Add(earningGroup.Key.MakePayment(paymentAmount));
+                }
+            }
+
+            foreach (var previousPaymentGroup in groupedPreviousPayments)
```
Comment (Pauljgraham) [inline src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs]: This second loop is doing a refund if the submission. Is it possible to break this into a function which returns the list of repayments

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    public class ProviderPaymentsCalculator : ICalculateProviderPayments
+    {
+        public List<PaymentEntity> CalculatePayments(
+            List<AdjustmentEntity> previousPayments, 
+            List<AdjustmentEntity> earnings)
+        {
+            var alreadyProcessedGroups = new HashSet<ProviderPaymentsGroup>();
+            var processedSubmissions = new HashSet<Guid>(
+                previousPayments.Select(x => x.SubmissionId)
+                    .Intersect(earnings.Select(y => y.SubmissionId)));
```
Comment (Pauljgraham) [inline src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs]: I'm not sure why we need to intersect here? Won't the new earnings submissions be different from the existing submissions

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+            var payments = new List<PaymentEntity>();
+
+            var groupedPreviousPayments = previousPayments.ToLookup(x => new ProviderPaymentsGroup(x));
+            var groupedEarnings = earnings
+                .Where(x => !processedSubmissions.Contains(x.SubmissionId))
+                .ToLookup(x => new ProviderPaymentsGroup(x));
+
+            foreach (var earningGroup in groupedEarnings)
```
Comment (Pauljgraham) [inline src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs]: Could we put this into a function which creates a list of new payments

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/ProviderAdjustmentsProcessor.cs [hunk 10]
```diff
@@ -36,29 +42,27 @@ public virtual void Process()
...
 
-                    ProcessAdjustmentsForProvider(provider, currentAdjustmentsResponse.Items, previousPaymentsResponse.Items, collectionPeriod.Period);
+            _logger.Info("Finished the Provider Adjustments Processor.");
+        }
 
-                    _logger.Info($"Finished processing adjustments for provider with ukprn {provider.Ukprn}.");
-                }
-            }
-            else
+        private void PopulateCollectionPeriod(IEnumerable<PaymentEntity> payments)
```
Comment (Pauljgraham) [inline src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/ProviderAdjustmentsProcessor.cs:58]: Does this method belong to the CalculatePayments Class.

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs (hunk 1 above)
Comment (carlbm) [reply src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs]: I vote for in the same file 👍

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs (hunk 2 above)
Comment (carlbm) [reply src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs]: Changed

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Domain/Extensions/ProviderPaymentsGroupExtensions.cs (hunk 3 above)
Comment (carlbm) [reply src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Domain/Extensions/ProviderPaymentsGroupExtensions.cs]: Changed

File: src/PeriodEnd/ProviderAdjustments/DeployProviderAdjustments/sql/dml/06 PeriodEnd.ProviderAdjustments.Populate.Reference.Current.dml.sql (hunk 4 above)
Comment (carlbm) [reply src/PeriodEnd/ProviderAdjustments/DeployProviderAdjustments/sql/dml/06 PeriodEnd.ProviderAdjustments.Populate.Reference.Current.dml.sql:20]: There are no cross-server joins in this query where before there was a join against a table in transient

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments.UnitTests/Sandbox.cs (hunk 5 above)
Comment (carlbm) [reply src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments.UnitTests/Sandbox.cs:10]: Interesting for future devs? Although it's not testing our code, our code uses what's being tested

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Domain/ProviderPaymentsGroup.cs (hunk 6 above)
Comment (carlbm) [reply src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Domain/ProviderPaymentsGroup.cs:9]: It's more data than an enum type, badly named maybe 😄

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs (hunk 7 above)
Comment (carlbm) [reply src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs]: Changed

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs (hunk 8 above)
Comment (carlbm) [reply src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs]: We want to process earnings without payments and payments without earnings. The intersect is the bit that we don't want to process

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs (hunk 9 above)
Comment (carlbm) [reply src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/Services/ProviderPaymentsCalculator.cs]: Changed

File: src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/ProviderAdjustmentsProcessor.cs (hunk 10 above)
Comment (carlbm) [reply src/PeriodEnd/ProviderAdjustments/SFA.DAS.Payments.Calc.ProviderAdjustments/ProviderAdjustmentsProcessor.cs:58]: It's not really doing anything around calculations and is more infrastructure related

**Review by Pauljgraham — APPROVED** [review]

**Pull request merged**